resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -506,6 +506,10 @@ private[spark] class MesosClusterScheduler(
       options ++= Seq("--class", desc.command.mainClass)
     }
 
+    desc.conf.getOption("spark.mesos.proxyUser").foreach { v =>
+      options ++= Seq("--proxy-user", v)
```

Comment:
This looks a little odd. How's a cluster mode app run in Mesos?

Basically what I want to know:
- which process starts the driver
- what user that process is running as, and which user will the driver process run as
- what kerberos credentials does it have and how are they managed

The gist is that running the Spark driver in client mode (which I think is how the driver in cluster mode is started eventually?) with a proxy user is a weird combination. It means the user code running in that driver has access to the credentials of the more privileged user - and could in its turn use those to run anything as any other user...

In comparison, YARN + cluster mode + proxy user starts the YARN application as the proxy user. So the user code, which only runs in a YARN container, has no access to the privileged credentials, which only exist in the launcher.

Reply:
@vanzin  

On dc/os the spark dcos cli which supports kerberos & keytab paths submits jobs directly to the 
mesos rest api at the mesos dispatcher side. The keytabs are stored on the DC/OS secret store before the job is launched and they are mounted on the container before container is launched.
The idea here is to store the keytab for the superuser on the secret store, so the spark driver which is eventually launched in client mode within the cluster, to login to kerberos and impersonate another user. The driver will start the the SparkJob's main as a proxy user (as usual) and will use the superuser credentials to impersonate the passed proxy user. 
The OS used by the container depends on the setup but that should be with minimal rights.
Right now DC/OS switched back to root, previously it used nobody but users can customize the image to add their users anywy. In containerized envs this add extra headaches.

Reply:
>  The driver will start the SparkJob's main as a proxy user (as usual) and will use the superuser credentials to impersonate the passed proxy user.

That's a big problem, because, as I said, that makes the super user credentials available to untrusted user code. How do you prevent the user's Spark app from using those credentials?

On YARN cluster mode the super user's credentials are never available to the user application. (On client mode they are, but really, if you're using `--proxy-user` in client mode you're missing the point, or I hope you know what you're doing.)

Reply:
Basically, you have a problem here you need to solve.

You either have to require kerberos creds on the launcher side, so you can upload DTs in cluster mode, or you need some level of separation between the code that launches the driver and the driver itself. The current system you have here is not secure at all - any user can just impersonate any other user, since they have access to the super user's credentials.

Reply:
There is a separation. The container launch and how DC/OS manages that is irrelevant to Spark. From the point that the dispatcher launches the driver and on ,then this is Spark's concern. The OS user DC/OS uses to launch the container is not related to the DC/OS user who submits the job or the Hadoop user. The DC/OS user who submits the spark job in cluster mode owns then superuser credentials and this is not leaked. Users here are not like Yarn users. 
If I use the DC/OS user who owns superuser's credentials all the way inside the container then
effectively I am ok. There is no requirement that impersonation should be a cluster wide characteristic. We are not solving the DC/OS impersonation issue here.
There are limitations btw under DC/OS but this is how it is right now check here:
https://docs.mesosphere.com/services/spark/2.3.0-2.2.1-2/limitations.

Reply:
I don't understand the separation.

You're starting `spark-submit` with `--proxy-user`, right? That *requires* the `spark-submit` process to know the credentials of the user that can impersonate another user.

If that `spark-submit` process is run in client mode, then it's running untrusted user code. Which means it also has access to those same super user credentials, and that's a security problem.

You have this in L530 below this code:

```
"spark.submit.deployMode", // this would be set to `cluster`, but we need client
```

What am I missing here?

Reply:
Yes this is how Spark runs on mesos so far. It is run in client mode, mesos dispatcher is just used to launch the driver in client mode. Its up to the DC/OS  to make sure that it runs things safely.
DC/OS does not use the spark-submit code it submits directly to the rest api of the dispatcher because spark submit is not convenient for submitting jobs outside the cluster. That submission is secured according to DC/OS security capabilities eg https access.
Also pure mesos does not have the infrastructure in place like (secret store or hdfs) to support this story, yarn has which assumes hdfs to manage secrets.
Now back to the scenario you describe let's assume I dont use the dc/os cli that submits to the rest api and avoids spark submit (for example I could do this from a node in my cluster, like yarn more or less).
Then I would use `spark-submit --proxy-user Y` but first I would have to log in my node as user X
who can impersonate other users and who can create a TGT for that puprose,
I would upload user's X ticket cache (which I can point to with KRB5CCNAME) on the cluster wide accessible ticket store. I would run spark submit with env var set to SPARK_USER=X which would launch via the dispatcher a container running the spark driver in client mode and that container would mount the TGT from the secret store. The OS user for the container would be X. 
In there the driver is launched in client mode impersonating user Y.

Note: check this PR: https://github.com/apache/spark/pull/20967,  a customer was trying to access remotely a cluster and spark-submit was not a good fit. It is hard to try unify all deployment envs,
and that is way DC/OS uses the cli and I think in the future scheduler code should be out of the project.

Reply:
I'm still confused about how submission works on Mesos in cluster mode. You mention a DC/OS CLI. Does that mean you're not using spark-submit?

The point I'm trying to make is that using `--proxy-user` in client mode in this context is a security issue. And I'm really uncomfortable with adding code in Spark that is basically a big security hole. You're basically giving up the idea of multiple users here, since by doing that any user can impersonate anyone else.

To comment on a few things:

> yarn has as it assumes hdfs to manage secrets.

No. Secrets (in this context, delegation tokens) are sent directly to YARN in the application request. No HDFS involved. That's done in `Client.setupSecurityToken` in the Spark code base.

The launcher has the TGT; the launcher creates DTs for services and attaches them to the YARN container.
work
So regardless of whether the user is running the app as himself, or impersonating another user, the application itself *does not* have access to the user's Kerberos credentials (not the password, not the keytab, not the TGT). It only has access to the DTs created by the launcher.

And that's why Spark-on-YARN is secure in this context.

> first I would have to log in my node as user X who can impersonate other users

I don't understand why do you have to impersonate at all here. If someone is launching an app, most of the time they want to launch the app as themselves. That's the "not impersonating" case in the above YARN scenario.

> I would upload user's X ticket cache (which I can point to with KRB5CCNAME) on the cluster 

Why? That's not even guaranteed to work (what if the TGT is bound to the machine that requested it?). That's why Spark-on-YARN creates DTs. That's what you should be giving the application.

Mesos in client mode does that AFAIK - because the user starting the application needs a proper kerberos login for things to work. But in cluster mode you need to figure out how to get the DTs to the application. And impersonation is not the way to do it for the reasons already explained.

So, going back to a previous suggestion of mine you have two ways of fixing this:

- require the launcher to have a kerberos login, and send DTs to the application. a.k.a. what Spark-on-YARN does.
- in the code that launches the driver on the Mesos side, create the DTs in a safe context (e.g. *not* as part of the spark-submit invocation) and provide them to the Spark driver using the `HADOOP_TOKEN_FILE_LOCATION` env var.

Sorry but what you have here won't work. If doing that, you might as well run all applications on your cluster as the super user - things will be just as secure.

Reply:
> I'm still confused about how submission works on Mesos in cluster mode. You mention a DC/OS > CLI. Does that mean you're not using spark-submit?

For dc/os there is a spark cli and spark submit is not used. People though use other ways to submit jobs.

> The point I'm trying to make is that using --proxy-user in client mode in this context is a security  > issue. And I'm really uncomfortable with adding code in Spark that is basically a big security > hole. You're basically giving up the idea of multiple users here, since by doing that any user can > impersonate anyone else.

If I use client mode and logged in as user X and have a TGT locally then why is it a security hole?
When the Spark main runs as a different user (proxy user) that user shouldnt have access to user's X stuff. Is not that true? I know DTs are more safe even if I steal them I cannot renew them, but if there is user isolation within the container then there is still a hole?
In Yarn the launcher in client mode still uses the same approach of uploading DTs?
In Yarn client mode the driver's main is run with the impersonated user right ? Then why that
user cannot access the TGT there?

Reply:
> If I use client mode and logged in as user X and have a TGT locally then why is it a security hole?
When the Spark main runs as a different user (proxy user) that user shouldnt have access to user's X stuff. Is not that true?

No, that is not true. That's why `HiveDelegationTokenProvider` can break out of the impersonated context, and use the original user to get a delegation token.

If that code can do it, any code can do it.

Think about it. User X's TGT is written in a file on the local file system. The file is only readable by user X, so for `spark-submit` to read it, it needs to run as user X. So, obviously, any code run by `spark-submit` can read that TGT - including the user code running as the "proxy user", since "proxy user" is not a concept that applies to the already running process, just how it authenticates to external services.

Reply:
> In Yarn client mode the driver's main is run with the impersonated user right. Then why that
user (proxy) cannot access the TGT there and why there is no hole then?

There is only impersonation if you use `--proxy-user`. If you don't, there is not. Still, the code that is running in the cluster (the AM and the executors) only sees DTs.

And yes, if you use impersonation in YARN client mode to run untrusted code, you're potentially exposing your kerberos credentials to malicious code. I mentioned this several times in my comments, that proxy users and client mode should not mix unless you really know what you're doing.

My problem here is that you're making spark-submit + proxy user + client mode the *official* way to run Spark on Mesos in cluster mode, and now you're basically exposing everyone to that security issue.

Reply:
Ok messy thought there was some guarantee... so in Yarn client mode how is the launcher isolated then?

Reply:
>I mentioned this several times in my comments, that proxy users and client mode should not mix > unless you really know what you're doing.

Ok I see now.

Reply:
> My problem here is that you're making spark-submit + proxy user + client mode the official way to > run Spark on Mesos in cluster mode, and now you're basically exposing everyone to that security > issue.
Yes because the assumption was client mode was safe. There is no warning about this especially for end users.

Reply:
> Yes because the assumption was client mode was safe. There is no warning about this 

Could probably use something in the documentation - warnings printed to logs are easily ignored. Still, there are legitimate uses for client mode + proxy user, but I don't think this is one of them.

Reply:
What are legitimate uses if it is not safe? Like knowing what your code does?

Reply:
> What are legitimate uses if it is not safe? Like knowing what your code does so its ok

Yes. For example you can run some trusted code as a less-privileged user, so that you don't accidentally do something stupid as a super user.

> (Mechanism level: Failed to find any Kerberos tgt)

That means you don't have any credentials (neither a tgt nor a dt). I don't know EMR so I don't know how to use it (with or without kerberos).

Reply:
>That means you don't have any credentials (neither a tgt nor a dt). I don't know EMR so I don't > know how to use it (with or without kerberos).

Yes that was the intention to check where Spark on Yarn code fails when there is no TGT. I am using it with kerberos.